pkg/installer/packages/manager.go

```diff
@@ -0,0 +1,300 @@
+	Install(ctx context.Context, url string) error
+	Remove(ctx context.Context, pkg string) error
+
+	InstallExperiment(ctx context.Context, url string) error
+	RemoveExperiment(ctx context.Context, pkg string) error
+	PromoteExperiment(ctx context.Context, pkg string) error
+
+	GarbageCollect(ctx context.Context) error
+}
+
+// managerImpl is the implementation of the package manager.
+type managerImpl struct {
+	m sync.Mutex
+
+	downloader   *downloader
+	repositories *repository.Repositories
+	configsDir   string
+	tmpDirPath   string
+	packagesPath string
+}
+
+// Options are the options for the package manager.
+type Options func(*options)
+
+type options struct {
+	registryAuth RegistryAuth
+	registry     string
+}
+
+func newOptions() *options {
+	return &options{
+		registryAuth: RegistryAuthDefault,
+		registry:     "",
+	}
+}
+
+// WithRegistryAuth sets the registry authentication method.
+func WithRegistryAuth(registryAuth RegistryAuth) Options {
+	return func(o *options) {
+		o.registryAuth = registryAuth
+	}
+}
+
+// WithRegistry sets the registry URL.
+func WithRegistry(registry string) Options {
+	return func(o *options) {
+		o.registry = registry
+	}
+}
+
+// NewManager returns a new Package Manager.
+func NewManager(opts ...Options) Manager {
+	o := newOptions()
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &managerImpl{
+		downloader:   newDownloader(http.DefaultClient, o.registryAuth, o.registry),
+		repositories: repository.NewRepositories(PackagesPath, LocksPack),
+		configsDir:   defaultConfigsDir,
+		tmpDirPath:   TmpDirPath,
+		packagesPath: PackagesPath,
+	}
+}
+
+// State returns the state of a package.
+func (m *managerImpl) State(pkg string) (repository.State, error) {
+	return m.repositories.GetPackageState(pkg)
+}
+
+// States returns the states of all packages.
+func (m *managerImpl) States() (map[string]repository.State, error) {
+	return m.repositories.GetState()
+}
+
+// Install installs or updates a package.
+func (m *managerImpl) Install(ctx context.Context, url string) error {
+	m.m.Lock()
+	defer m.m.Unlock()
+	err := utils.CheckAvailableDiskSpace(mininumDiskSpace, m.packagesPath)
+	if err != nil {
+		return fmt.Errorf("not enough disk space: %w", err)
+	}
+
+	pkg, err := m.downloader.Download(ctx, url)
+	if err != nil {
+		return fmt.Errorf("could not download package: %w", err)
+	}
+	tmpDir, err := os.MkdirTemp(m.tmpDirPath, fmt.Sprintf("install-stable-%s-*", pkg.Name)) // * is replaced by a random string
+	if err != nil {
+		return fmt.Errorf("could not create temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	configDir := filepath.Join(m.configsDir, pkg.Name)
+	err = extractPackageLayers(pkg.Image, configDir, tmpDir)
+	if err != nil {
+		return fmt.Errorf("could not extract package layers: %w", err)
+	}
+	err = m.repositories.Create(ctx, pkg.Name, pkg.Version, tmpDir)
+	if err != nil {
+		return fmt.Errorf("could not create repository: %w", err)
+	}
+	return m.setupUnits(ctx, pkg.Name)
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 Nitpick:** ‏It's not just units that are setup, we should probably rename to `setupPackage` or just `setup` for clarity

Reply:
Let's do as a followup (probably by burning a lot of this entirely...)

Reply:
was previously the same, will address in a followup